### ISSUE TITLE
[no-inferrable-types] flags types that aren't inferable

### ISSUE THREAD

**Issue by dleavitt** [typescript-eslint/typescript-eslint#2184]
In a way that conflicts with the `no-unsafe-assignment` rule.

**Repro**

```TS
// **no-inferrable-types** is triggered
const logOut = useCallback((redirectPath: string = "/users/sign_in") => {
  window.location.href = redirectPath
}, [])
```

Occurs with default settings for `@typescript-eslint/eslint-plugin`, specifically:
```JSON
{
  "rules": {
    "@typescript-eslint/no-inferrable-types": "error",
    "@typescript-eslint/no-unsafe-assignment": "error",
  }
}
```

**Expected Result**
`no-inferrable-types` should **not** be triggered, as the type of `redirectPath` cannot be inferred.

**Actual Result**
`no-inferrable-types` **is** triggered.

**Additional Info**

`redirectPath` cannot be inferred to be a string in the above context. Removing the type annotation will trigger cause the `no-unsafe-assignment` rule to fail:

```TS
const logOut = useCallback((redirectPath = "/users/sign_in") => {
  // **no-unsafe-assignment** is triggered
  window.location.href = redirectPath
}, [])
```
So the `no-unsafe-assignment` rule knows that the type is not inferable, but the `no-inferrable-types` thinks that it is.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.1.0` |
| `@typescript-eslint/parser`        | `3.1.0` |
| `TypeScript`                       | `3.9.5` |
| `ESLint`                           | `7.2.0` |
| `node`                             | `13.11.0` |
| `npm`                              | `6.13.7` |

**Comment by bradzacher** [comment]
This incorrect parameter typing inference is due to how the react types define the function type: `TCallback extends (...args: any[]) => any`.

This causes TS to implicitly type the arguments as `any`.

There's nothing we can do about this from the perspective of this rule.
You'll have to use a disable comment here as it's an exception to the rule.

Consider submitting an issue against the react types.

**Comment by dleavitt** [comment]
Makes sense; thanks for the clarification. Should the `no-inferrable-types` rule avoid flagging cases like this where the explicit type annotation narrows the parameter type relative to what's in the type signature of the method being called? Is that feasible/sensical?

**Comment by bradzacher** [comment]
> Is that feasible?

No.  
This rule does not use type information, so it cannot detect this.  

TBH I don't think it's even possible to detect this, even with type information - it's a complex interaction between the generic type and the value.
